java/vector/src/main/java/org/apache/arrow/vector/VectorUnloader.java

```diff
@@ -69,9 +85,79 @@ private void appendNodes(FieldVector vector, List<ArrowFieldNode> nodes, List<Ar
           vector.getField(), vector.getClass().getSimpleName(), fieldBuffers));
     }
     buffers.addAll(fieldBuffers);
+
     for (FieldVector child : vector.getChildrenFromFields()) {
       appendNodes(child, nodes, buffers);
     }
   }
 
+  // translate dictionary fields from in-memory format to message format
+  // add dictionary ids, change field types to dictionary type instead of index type
+  private void updateSchemaAndUnloadDictionaries() {
+    Map<Dictionary, ArrowDictionaryBatch> dictionaries = new HashMap<>();
+    Map<Dictionary, Long> dictionaryIds = new HashMap<>();
+
+    // go through once and collect any existing dictionary ids so that we don't duplicate them
+    for (FieldVector vector: vectors) {
+      if (vector instanceof DictionaryVector) {
+        Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+        dictionaryIds.put(dictionary, dictionary.getId());
+      }
+    }
+
+    // now generate ids for any dictionaries that haven't defined them
+    long nextDictionaryId = 0;
+    for (Entry<Dictionary, Long> entry: dictionaryIds.entrySet()) {
+      if (entry.getValue() == null) {
+        while (dictionaryIds.values().contains(nextDictionaryId)) {
+          nextDictionaryId++;
+        }
+        dictionaryIds.put(entry.getKey(), nextDictionaryId);
+      }
+    }
+
+    // go through again to add dictionary id to the schema fields and to unload the dictionary batches
+    for (FieldVector vector: vectors) {
+      if (vector instanceof DictionaryVector) {
+        Dictionary dictionary = ((DictionaryVector) vector).getDictionary();
+        long dictionaryId = dictionaryIds.get(dictionary);
```

Comment:
If I add another method to FieldVector, won't I need to change the codegen packages? In the previous dictionary encoding PR, @wesm asked me to keep the dictionary vector as a 'virtual type', and not to modify the codegen packages. Thus the only way to check for a dictionary vector is to use `instanceof`. I'm fine doing it either way.